### ISSUE TITLE
Error with TELNET no enable

### ISSUE THREAD

**Issue by pauloj95** [rconfig/rconfig#82]
Hi,

**UPDATE:**
So the actual problem is:

I cant download config/snippets via TELNET:
In log i can see the Success connected:

2018-03-12 16:26:33 INFO The /lib/ajaxHandlers/ajaxConfigDevice.php script was run manually invoked with Router ID: 4 & Snippet ID: 1
2018-03-12 16:26:36 CONN Success: Connected to Router (10.x.xxx.x) for Router ID 4 - in (File: /lib/ajaxHandlers/ajaxConfigDevice.php)

But then nothing happens...no download just no result like the image i've sent...
![error](https://user-images.githubusercontent.com/37109548/37296861-143c4e7e-2614-11e8-96a9-2d763553772e.PNG)

I have tested telnet via shell and works like charm...

Thanks.
PJesus

**Comment by pauloj95** [comment]
UPDATE: i'm having this problem only via TELNET... i cant download configurations too...only can to ssh devices...

**Comment by pauloj95** [comment]
**UPDATE**
So the actual problem is:

I cant download config/snippets via TELNET:
In log i can see the **Success** connected:

2018-03-12 16:26:33 INFO The /lib/ajaxHandlers/ajaxConfigDevice.php script was run manually invoked with Router ID: 4 & Snippet ID: 1 
2018-03-12 16:26:36 CONN Success: Connected to Router (10.x.xxx.x) for Router ID 4 - in (File: /lib/ajaxHandlers/ajaxConfigDevice.php)


But then nothing happens...no download just no result like the image i've sent...
I have tested telnet via shell and works like charm...

Thanks.
PJesus

**Comment by pauloj95** [comment]
Looks like there is more people with this problem:

https://github.com/rconfig/rConfig-templates/issues/4

Thanks a lot,

**Comment by ayad99** [comment]
Hi

We had the same issue. I have checked all the points suggested here (and in kb2001) by rConfig team. But it simply does not work with Telnet. Interestingly I used another tool and it picked up the config dump fine using the same Telnet credentials. So this is not a device config issue, or device credential issue. And this is not a rConfig issue as the settings are pretty straight forward. SSH works fine though in rConfig. It seems like there is a hidden bug in rConfig.

I understand that the rConfig tool is free, but the support is paid. But the way rConfig team is asking to buy support to identify this hidden issue appears bit unethical. It would have been better if rConfig was not offered as a free tool and rather had a paid subscription in the first place. At least we would all know what to expect.

**Comment by rconfig** [comment]
@ayad99 Please read this article. Your accusation of being unethical is ironic, considering you just spammed my issues form with the same message on multiple threads. 

https://medium.com/@d4nyll/the-open-source-community-have-no-place-for-disrespect-70c85d473332 

@pauloj95 If you post an output from KB2001, i will aim to see if I can pinpoint your issue.

**Comment by ayad99** [comment]
Hi @rconfig Stephen

I never accused anyone. I just said it "appears" to be so. My apologies if it was offensive. And I did not spam the issues form. I copied the same comment on all the issues related to Telnet not working, since there were multiple.

**Comment by wajahatraza** [comment]
@rconfig I am facing the same issue [@pauloj95 referred it]. I do checked using KB2001 and following was the output:
Following is the output when I am using TELNET:

The /home/rconfig/lib/showCmdScript.php script was run from a manual invocation on a shell with Task ID:456078
19 March 2018 04:49:10 PM
Array
(
    [id] => 8
    [deviceName] => ABC
    [deviceIpAddr] => 1.1.1.1
    [deviceEnablePrompt] =>
    [devicePrompt] => SW#
    [deviceUsername] => abc
    [devicePassword] => abc@wsx
    [deviceEnablePassword] =>
    [templateId] => 1
    [nodeCatId] => 1
)
Success: Connected to switch (1.1.1.1) for taskID 456078 - in (File: /home/rconfig/lib/showCmdScript.php)
19 March 2018 04:49:13 PM
sh runterminal length 0
WB-SW#\r\n------------------------------------------------------\r\n
WB-SW#19 March 2018 04:49:13 PM
Array
(
)
PHP Notice:  Undefined index: saveConfig in /home/rconfig/lib/showCmdScript.php on line 265

Notice: Undefined index: saveConfig in /home/rconfig/lib/showCmdScript.php on line 265
The owner permisions for directory /home/rconfig/data/ were reset to owner apache because script was run interactively


Following is the output when I am using the SSH for same device as above:

The /home/rconfig/lib/showCmdScript.php script was run from a manual invocation on a shell with Task ID:456078
19 March 2018 04:50:11 PM
Array
(
    [id] => 8
    [deviceName] => abc
    [deviceIpAddr] => 1.1.1.1
    [deviceEnablePrompt] =>
    [devicePrompt] => SW#
    [deviceUsername] => abc
    [devicePassword] => abc@wsx
    [deviceEnablePassword] =>
    [templateId] => 3
    [nodeCatId] => 1
)
19 March 2018 04:50:11 PM
sh runSuccess: Connected via SSH to switch(1.1.1.1) for command (sh run) for taskID 456078 - in (File: /home/rconfig/lib/showCmdScript.php)
19 March 2018 04:50:14 PM
Array
(
    [0] => Building configuration...
    [1] =>
...) [It is running completely file using SSH]

**Comment by rconfig** [comment]
@wajahatraza your issue appears to be straightforward. Check your template file. The error is;
Notice: Undefined index: saveConfig in /home/rconfig/lib/showCmdScript.php on line 265

Your template needs to follow this standard, and is missing the line 'saveConfig' per below. Make sure the 'saveConfig' line in contained within the 'config: ' section. 
https://github.com/rconfig/rConfig-templates/blob/master/base/base.yml
saveConfig: "wr mem"

And always read the help files :)

https://www.rconfig.com/help/devices/templates

Let me know if this works or not

**Comment by wajahatraza** [comment]
@rconfig Thanks for your response, I do have commented out "wr mem" because I don't want to save the configuration. I have disabled it in SSH plugin too but it is working fine. Even before disabling the wr mem, it was not working.

**Comment by rconfig** [comment]
ok, then maybe I need to update help docs. Dont remove this
saveConfig: "wr mem"
leave it blank instead
saveConfig: " "

It wont invoke a save

**Comment by wajahatraza** [comment]
I have left it blank too but still telnet is not working. As mentioned above, I have commented out the same line in SSH and yet it works fine. I think that rConfig is able to access the device and execute the command properly but unable to capture output. Also to mention that I have checked the device credentials/config/firewall settings but still TELNET output isn't working.

**Comment by rconfig** [comment]
if its left blank, run the debug again and post the output

**Comment by wajahatraza** [comment]
@rconfig I haved configured saveConfig: " " in the connection template and following is the output:

The /home/rconfig/lib/showCmdScript.php script was run from a manual invocation                                                                                         on a shell with Task ID:743127
20 March 2018 02:52:11 PM
Array
(
    [id] => 8
    [deviceName] => ABC
    [deviceIpAddr] => 1.1.1.1
    [deviceEnablePrompt] =>
    [devicePrompt] => SW#
    [deviceUsername] => abc
    [devicePassword] => abc@wsx
    [deviceEnablePassword] =>
    [templateId] => 1
    [nodeCatId] => 1
)
Success: Connected to Switch (1.1.1.1) for taskID 743127 - in (File: /home/rconfig/lib/showCmdScript.php)
20 March 2018 02:52:14 PM
show versionterminal length 0
SW#\r\n------------------------------------------------------\r\n
SW#20 March 2018 02:52:14 PM
Array
(
)
The owner permisions for directory /home/rconfig/data/ were reset to owner apache because script was run interactively

**Comment by rconfig** [comment]
Ok, thats better. It connects. But as you see nothing is returned. What category is this device set for? And what commands are configured for this category?

You can run a deep debug but opening the classes/connection2.class.php file and uncommenting line 196. Change this;
 //                     var_dump($this->_data);
to this
var_dump($this->_data);

Run the debug again. Also, what code version is this device? and are any devices working for you at all? or is just this one failing?

**Comment by wajahatraza** [comment]
Thanks for your response. Category set for this device is Switch and we have configured "show version" command in this category.
This device model is Cisco 2960. Please note that irrespective of device model and type, TELNET output is not visible, only SSH is working.
I will update you the result after editing the classes/connection2.class.php file.

**Comment by wajahatraza** [comment]
Thanks for your response. Category set for this device is Switch and we have configured "show version" command in this category.
This device model is Cisco 2960. Please note that irrespective of device model and type, TELNET output is not visible, only SSH is working.
I will update you the result after editing the classes/connection2.class.php file.

**Comment by wajahatraza** [comment]
Please find attached output file with deep debug as suggested. I have used "sh int desc" command in this diagnosis as the file with the output of show version command was large.
[TELNET Output.txt](https://github.com/rconfig/rconfig/files/1829899/TELNET.Output.txt)

**Comment by rconfig** [comment]
ok, so this output is good. It actually looks like you got the config. You ran a 'sh int desc' on the switch. And it ran successfully and was returned. Was this config output available in rConfig? 
I also note the switch prompt is WB-SW#. Early i saw it was configured as SW1. rConfig needs to be configured with an exact match,

**Comment by wajahatraza** [comment]
This is where the issue is, the output is not available in rConfig. It shows blank window in the ouput and I have also checked the output file from the CLI and it also shows blank file there. 
Regarding the switch name, please leave it as I changed the actual switch name due to privacy/security issues. I have configured it correctly in the rConfig and it worked fine when using telnet.
Following is the output from /data/../switch directory:
-r--r--r-- 1 apache root      0 Mar 20 19:15 shintdesc-1915.txt
-r--r--r-- 1 apache root      0 Mar 20 19:16 shintdesc-1916.txt
-r--r--r-- 1 apache root      0 Mar 20 19:17 shintdesc-1917.txt
-r--r--r-- 1 apache root      0 Mar 20 23:32 shintdesc-2332.txt
-r--r--r-- 1 apache apache    0 Mar 20 23:35 shintdesc-2335.txt
-r--r--r-- 1 apache root      0 Mar 20 14:52 showversion-1452.txt

**Comment by pauloj95** [comment]
@rconfig here is my result of running kb2001

The /home/rconfig/lib/showCmdScript.php script was run from a manual invocation on a shell with Task ID:652168
23 March 2018 04:28:47 PM
Array
(
    [id] => 67
    [deviceName] => PTLavWap01
    [deviceIpAddr] => 10.5.253.8
    [deviceEnablePrompt] => PTLV1P10WAP07>
    [devicePrompt] =>  PTLV1P10WAP07#
    [deviceUsername] => admin
    [devicePassword] => **********************
    [deviceEnablePassword] => **********************
    [templateId] => 2
    [nodeCatId] => 12
)
PHP Notice:  Undefined index: sshInteractive in /home/rconfig/lib/showCmdScript.php on line 147

Notice: Undefined index: sshInteractive in /home/rconfig/lib/showCmdScript.php on line 147
Failure: Unable to connect to PTLavWap01 - 10.5.253.8 when running taskID 652168 - in  Error:(File: /home/rconfig/lib/showCmdScript.php)
The owner permisions for directory /home/rconfig/data/ were reset to owner apache because script was run interactively

**Comment by wajahatraza** [comment]
Hi @rconfig Stephen,
Have you been able to found any bug or been able replicate this issue? Just to clarify that we are getting the output using deep debug mode on CLI but not getting any output on rConfig web console or in ../data/.. directory.

**Comment by rconfig** [comment]
can you open the template for this device, and remove the line where the key is 'sshInteractive'

try this again

**Comment by wajahatraza** [comment]
I couldn't find any line where they key is 'sshInteractive'. I have checked in the connection template, showCmdScript.php and connection2.class.php. Can you please specify which file do I need to check?

**Comment by rconfig** [comment]
ok, dowload the 3.8.7 release from rconfig.com. replace the connection2.class.php from the 3.8.7 download, with the file you have installed. Test again

**Comment by wajahatraza** [comment]
I already have version 3.8.7 installed. Following is the version from the console's footer:
rConfig Version 3.8.7   © rConfig 2015 - 2018

**Comment by rconfig** [comment]
that cannot be true. If your connection2.class.php is throwing the 'sshInteractive' error. That can ONLY mean, you acquired the connection2.class.php file from this github repo. Thats a newer version, and per documentation on rconfig.com only 3.8.7 is supported. If you do as I ask above, the 3.8.7 zip from rconfig.com has the correct files.

For posterity, i will update this repo later today to remove the sshInteractive code.

Also, the installation documentation only allows for installations from rconfig.com... not github.

Thanks

**Comment by wajahatraza** [comment]
I guess you are confusing my comment with @pauloj95 comment. I am not getting the sshInteractive error. As I mentioned above that the connection is successful, and the script produced the output on the CLI when I ran it. However this output is not captured by rConfig and the file it saves is also empty.

Further more, I installed the rConfig 3.8.7 from rConfig.com website (not github) using the below command as mentioned in the Installation guide.

cd /home
curl -O http://www.rconfig.com/downloads/scripts/install_rConfig.sh -A "Mozilla"
chmod +x install_rConfig.sh
./install_rConfig.sh
